Error while checking build/O2/fullCI for fc2a17a575f2551f163cb60f605dca6a46ed4fd2 at 2024-02-14 17:15:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:45:44: error: template argument 2 is invalid
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:45:44: error: template argument 5 is invalid
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:46:16: error: reference to 'Input' is ambiguous
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:47:14: error: reference to 'Input' is ambiguous
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:45:46: error: scalar object 'InputMap' requires one element in initializer
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:59:20: error: template argument 1 is invalid
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:59:20: error: template argument 2 is invalid
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:61:38: error: parse error in template argument list
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:61:53: error: no matching function for call to 'o2::RangeTokenizer::tokenize<<expression error> >(std::__cxx11::basic_string<char>&, defineDataProcessing(const o2::framework::ConfigContext&)::<lambda(const std::string&)>)'
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:61:53: error: template argument 1 is invalid
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:65:34: error: reference to 'Input' is ambiguous
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:65:34: error: 'Input' has not been declared
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:66:33: error: request for member 'begin' in 'inputTypes', which is of non-class type 'int'
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:66:53: error: request for member 'end' in 'inputTypes', which is of non-class type 'int'
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:66:80: error: request for member 'end' in 'inputTypes', which is of non-class type 'int'
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:69:17: error: reference to 'Input' is ambiguous
/sw/SOURCES/O2/12707-slc8_x86-64/0/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx:76:17: error: reference to 'Input' is ambiguous
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12707/fc2a17a575f2551f163cb60f605dca6a46ed4fd2/build_O2_fullCI/pretty.html).

--- a/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx
+++ b/Detectors/TPC/workflow/src/FileReaderWorkflow.cxx
@@ -38,13 +38,13 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
 
 using namespace o2::framework;
 
-enum struct Input { Clusters,
-                    Tracks
+enum struct InputType { Clusters,
+                        Tracks
 };
 
-const std::unordered_map<std::string, Input> InputMap{
-  {"clusters", Input::Clusters},
-  {"tracks", Input::Tracks}};
+const std::unordered_map<std::string, InputType> InputMap{
+  {"clusters", InputType::Clusters},
+  {"tracks", InputType::Tracks}};
 
 /// MC info is processed by default, disabled by using command line option `--disable-mc`
 ///
@@ -56,24 +56,24 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
   auto inputType = cfgc.options().get<std::string>("input-type");
   bool doMC = not cfgc.options().get<bool>("disable-mc");
 
-  std::vector<Input> inputTypes;
+  std::vector<InputType> inputTypes;
   try {
-    inputTypes = o2::RangeTokenizer::tokenize<Input>(inputType, [](std::string const& token) { return InputMap.at(token); });
+    inputTypes = o2::RangeTokenizer::tokenize<InputType>(inputType, [](std::string const& token) { return InputMap.at(token); });
   } catch (std::out_of_range&) {
     throw std::invalid_argument(std::string("invalid input type: ") + inputType);
   }
-  auto isEnabled = [&inputTypes](Input type) {
+  auto isEnabled = [&inputTypes](InputType type) {
     return std::find(inputTypes.begin(), inputTypes.end(), type) != inputTypes.end();
   };
 
-  if (isEnabled(Input::Clusters)) {
+  if (isEnabled(InputType::Clusters)) {
     specs.emplace_back(o2::tpc::getClusterReaderSpec(doMC));
     if (!getenv("DPL_DISABLE_TPC_TRIGGER_READER") || atoi(getenv("DPL_DISABLE_TPC_TRIGGER_READER")) != 1) {
       specs.emplace_back(o2::tpc::getTPCTriggerReaderSpec());
     }
   }
 
-  if (isEnabled(Input::Tracks)) {
+  if (isEnabled(InputType::Tracks)) {
 
     specs.push_back(o2::tpc::getTPCTrackReaderSpec(doMC));
   }

--- a/Framework/Core/include/Framework/AlgorithmSpec.h
+++ b/Framework/Core/include/Framework/AlgorithmSpec.h
@@ -84,6 +84,33 @@ struct AlgorithmPlugin {
   virtual AlgorithmSpec create() = 0;
 };
 
+template <size_t N>
+struct StringLiteral {
+  constexpr StringLiteral(const char (&str)[N])
+  {
+    std::copy_n(str, N, value);
+  }
+
+  char value[N];
+};
+
+// Allow fetching inputs from the context using a string literal.
+template <StringLiteral lit, typename T>
+struct Input {
+  // The size of the string is available as a constant expression.
+  static constexpr auto size = sizeof(lit.value);
+  // and so is the string's content.
+  static constexpr auto contents = lit.value;
+  ProcessingContext& ctx;
+  Input(ProcessingContext& c) : ctx(c)
+  {
+  }
+  operator T const&() const
+  {
+    return ctx.inputs().template get<T>(lit.value);
+  }
+};
+
 template <typename T, typename S = std::void_t<>>
 struct ContextElementTraits {
   static decltype(auto) get(ProcessingContext& ctx)
@@ -137,6 +164,22 @@ struct ContextElementTraits<ProcessingContext> {
   }
 };
 
+template <StringLiteral L, typename T>
+struct ContextElementTraits<Input<L, T> const> {
+  static Input<L, T> get(ProcessingContext& ctx)
+  {
+    return Input<L, T>{ctx};
+  }
+};
+
+template <StringLiteral L, typename T>
+struct ContextElementTraits<Input<L, T>> {
+  static Input<L, T> get(ProcessingContext& ctx)
+  {
+    static_assert(always_static_assert_v<Input<L, T>>, "Should be Input<L, T> const&");
+  }
+};
+
 template <typename... CONTEXTELEMENT>
 AlgorithmSpec::ProcessCallback adaptStatelessF(std::function<void(CONTEXTELEMENT&...)> callback)
 {
@@ -177,6 +220,11 @@ AlgorithmSpec::ProcessCallback adaptStatelessP(R (*callback)(ARGS...))
 ///   inputs.get<int>("someInt");
 /// }}
 ///
+/// and if you have C++20 enabled you can also do:
+///
+/// AlgorithmSpec{[](Input<"someInt", int> someInt){
+///  someInt.value; // do something with the inputs
+/// }
 /// Notice you can specify in any order any of InputRecord, DataAllocator,
 /// ConfigParamRegistry or any of the services which are usually hanging
 /// from the ServiceRegistry, e.g. ControlService.

